File: 剑指Offer/面试题56 - I. 数组中数字出现的次数.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var singleNumbers = function (nums) {
  let r = nums.reduce((s, x) => s ^ x, 0);
  let d = 1;
  while ((d & r) === 0) d <<= 1;
  let a = 0, b = 0;
  for (let x of nums) {
    if (x & d) a ^= x;
    else b ^= x;
  }
  return [a, b];
};

File: 剑指Offer/面试题56 - I. 数组中数字出现的次数.rs
impl Solution {
  pub fn single_numbers(nums: Vec<i32>) -> Vec<i32> {
      let mut r=0;
      for x in &nums{
          r^=x;
      }
      let mut d=1;
      while (d&r)==0{
          d<<=1;
      }
      let mut a=0;
      let mut b=0;
      for x in &nums{
          if x&d==0{
              a^=x;
          }
          else{
              b^=x;
          }
      }
      return vec![a,b];
  }
}